### PR TITLE
AvgPool: Ensure all cells are valid in ceil mode

### LINKED ISSUE #36977: Under `ceil_mode=True` for `AvgPooling2d`, Pytorch fails in calculating pooling output shape as expected, and gets `NaN` results.
## 🐛 Bug

**Under `ceil_mode=True` for `AvgPooling2d`, Pytorch fails in calculating pooling output shape as expected, and gets `NaN` results, no matter executing with CPU version or GPU version.**

**Taking Pytorch with CPU version as an example, given an input with the shape (1, 16, 4, 4), I get an unexpected output under layer the `AvgPool2d` , which contains `NaN`(Not a Number) results**，as shown in the following picture.

avg-fig2

 In this case, the parameters of `AvgPool2d` are as follows:

```python
{
  'kernel_size': (1, 2), 
 'padding': (0, 1), 
 'stride': (2, 2),
 'ceil_mode': True, 
 'count_include_pad': True}
```

To investigate the root cause of such `NaN` results, I conduct a code tracing in the source code of Pytorch when executing `AvgPool2d`. Relevant code can be found in`aten/src/ATen/native/Pool.h`  and `aten/src/ATen/native/AveragePool2d.cpp`. 

Firstly, Pytorch invokes the function `pooling_output_shape_pad_lr` in `aten/src/ATen/native/AveragePool2d.cpp line 129-130`  to calculate to pooling output shape. 

avg-fig3

Then Pytorch passes the two parameters `outputHeight` and `outputWidth` to function `avg_pool2d_out_frame` in `aten/src/ATen/native/AveragePool2d.cpp (near line 14)` to calculate the average pooling for each pool unit. 

The implementation of `pooling_output_shape_pad_lr` is defined in `aten/src/ATen/native/Pool.h (near line 25)` . See the following picture.

if_pad_l

The parameter `ceil_mode=True` indicates Pytorch uses ceil value instead of floor value when getting an output (refer to the official  [website](https://pytorch.org/docs/stable/nn.html?highlight=avgpool2d#torch.nn.AvgPool2d)). For pooling calculations,  the `ceil_mode=True` may make the last pool unit start outside the image. From the above code, we know that Pytorch tries to avoid such condition  by executing `--outputsize`. This indeed works, but is only be triggered when using padding on this dimension (see`if(pad_l) line 31`). In other words, if there is no padding  (`pad_l=0`) on this dimension, the outputsize will not be reduced by 1.

As a result, passing the output shape (also called `out-of-range` output shape) to `avg_pool2d_out_frame`, it will cause the last pool unit to be outside the image, as shown in the following example.

avg-fig1

To be specific, an `out-of-range` output shape will generate a `hstart`(or `wstart`) whose value is greater or equal to `hend`（or `wend`), thus making the `pool_size=(hend-hstart)*(wstart-wend)` be zero or a negative value,  as shown between the `line 46-83` in  `aten/src/ATen/native/AveragePool2d.cpp`.  

hstart=hend

In particular, just as our example, when `hstart=hend` (or `wstart=wend` ), it will make `poolsize = 0`,`divide_factor=0` and`sum = 0`. Consequently,  the corresponding average pooling unit is calculated with  0 divided by 0, thus triggering `NaN`. Actually, the premise of function `avg_pool2d_out_frame` is that `hstart ` and `wstart ` are always smaller than `hend` and `wend`, and this needs to be guaranteed by the previous function `pooling_output_shape_pad_lr` in `pool.h`. But due to `if (pad_l)`, it only executes the relevant judgments and reductions when using padding on this dimension.

In addition, I also try multiple sets of other different parameters. They all trigger the `NaN` problem 
 on both CPU and GPU version, because of the same reason (`hstart=hend`), as shown in the following table.

| Parameters for `AvgPool2D`                                   | Input Shape     | Output Shape    |
| ------------------------------------------------------------ | --------------- | --------------- |
| {'ceil_mode': True, 'count_include_pad': True, 'kernel_size': (1, 4), 'padding': (0, 1), 'stride': (2, 2)} | (1, 16, 32, 32) | (1, 16, 17, 16) |
| {'ceil_mode': True, 'count_include_pad': True, 'kernel_size': 2, 'padding': (0, 1), 'stride': (3, 3)} | (1, 16, 9, 9)   | (1, 16, 4, 4)   |
| {'ceil_mode': True, 'count_include_pad': True, 'kernel_size': (1, 3), 'padding': (0, 1), 'stride': (4, 4)} | （1,16,4,4）    | (1, 16, 2, 2)   |

### Key insight

**Under the `ceil_mode=True`, the statement `if (pad_l)`  in function `pooling_output_shape_pad_lr` (see `aten/src/ATen/native/Pool.h` ) cannot reduce the `out-of-range` outputshape, making some pools are taken outside the input image.  Such situations occur in both CPU version and GPU version. Even worse, the special case when `hstart=hend` or `wstart=wend` will trigger the `NaN` output. Such `NaN` output in pooling operations will confuse users, and also will affect the subsequent calculation in the model, causing unexpected results.**

## To Reproduce

1. Just run the following code.

### Code to reproduce 

```
import numpy as np
import torch
   
kwargs = {
	'ceil_mode': True, 
	'count_include_pad': True, 
	'kernel_size': (1, 2), 
	'padding': (0, 1), 
	'stride': (2, 2)
	}  
input = (10 * np.random.randn(1,16,4,4)).astype(np.float32)    
layer = torch.nn.AvgPool2d(**kwargs)
modelinput=torch.from_numpy(input)
output = layer(modelinput)
ytrue=output.detach().numpy()
nanresult=np.isnan(ytrue).any()

print(nanresult)
```

## Expected behavior

The output is a matrix of constants, at least no 'Nan'

## Environment
Using /pytorch/torch/utils/collect_env.py to collect the environment.
```

Collecting environment information...
PyTorch version: 1.6.0a0+6ba734b
Is debug build: No
CUDA used to build PyTorch: Could not collect

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: TITAN V
GPU 1: TITAN V

Nvidia driver version: 430.26
cuDNN version: /usr/local/cuda-9.0/lib64/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.4.0+cu92
[pip] torchvision==0.5.0+cu92
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.0                      166  
[conda] mkl-include               2020.0                      166  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.15           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] numpy                     1.18.1           py36h4f9e942_0  
[conda] numpy-base                1.18.1           py36hde5b4d6_1  
[conda] torch                     1.6.0a0+6ba734b          pypi_0    pypi

```

 - PyTorch Version (e.g., 1.0): 1.6.0a0+6ba734b(It is a CPU version, I compiled from Pytorch(master branch) source code ); 1.4.0-cu92 ( It is a GPU version, I used ` pip install torch==1.4.0+cu92 torchvision==0.5.0+cu92 -f https://download.pytorch.org/whl/torch_stable.html ` command to install )
 - OS (e.g., Linux): Linux Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source):( Just follow the guide in `https://github.com/pytorch/pytorch`, using `python setup.py install `)
 - Python version: 3.6.10
 - CUDA/cuDNN version: 9.2
 - GPU models and configuration:TITAN V
 - Any other relevant information:

cc @ezyang @gchanan @zou3519 @albanD @mruberry

### PR DESCRIPTION
**Pull request by peterbell10**
Closes #36977

This avoid the division by zero that was causing NaNs to appear in the output. `AvgPooling2d` and `AvgPooling3d` both had this issue on CPU and CUDA.

### CODE REVIEW

**Comment by ezyang** [comment]
Needs benchmarks before and after

**Review by ezyang — APPROVED** [review]
Logic looks pretty reasonable. Benchmarking should show a perf hit but hopefully it is not too bad.

**Comment by peterbell10** [comment]
I've run the long benchmarks for AvgPool2d and AvgPool3d. Cuda benchmarks show no measurable difference, cpu benchmark shows up to a 1% slowdown for `AvgPool2d` and 2% for `AvgPool3d`.
